pennylane/math/single_dispatch.py

```diff
@@ -0,0 +1,327 @@
+# limitations under the License.
+"""Autoray registrations"""
+# pylint:disable=protected-access,import-outside-toplevel,wrong-import-position
+from importlib import import_module
+import numbers
+
+import autoray as ar
+import numpy as np
+
+
+def _i(name):
+    """Convenience function to import PennyLane
+    interfaces via a string pattern"""
+    if name == "tf":
+        return import_module("tensorflow")
+
+    if name == "qml":
+        return import_module("pennylane")
+
+    return import_module(name)
+
+
+# -------------------------------- NumPy --------------------------------- #
+from scipy.linalg import block_diag as _scipy_block_diag
+
+ar.register_function("numpy", "flatten", lambda x: x.flatten())
+ar.register_function("numpy", "coerce", lambda x: x)
+ar.register_function("numpy", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("builtins", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("numpy", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _scatter_element_add_numpy(tensor, index, value):
+    tensor[tuple(index)] += value
+    return tensor
+
+
+ar.register_function("numpy", "scatter_element_add", _scatter_element_add_numpy)
+
+
+# -------------------------------- Autograd --------------------------------- #
+
+
+# When autoray inspects PennyLane NumPy tensors, they will be associated with
+# the 'pennylane' module, and not autograd. Set an alias so it understands this is
+# simply autograd.
+ar.autoray._BACKEND_ALIASES["pennylane"] = "autograd"
+
+# When dispatching to autograd, ensure that autoray will instead call
+# qml.numpy rather than autograd.numpy, to take into account our autograd modification.
+ar.autoray._MODULE_ALIASES["autograd"] = "pennylane.numpy"
+
+
+ar.register_function("autograd", "flatten", lambda x: x.flatten())
+ar.register_function("autograd", "coerce", lambda x: x)
+ar.register_function("autograd", "block_diag", lambda x: _scipy_block_diag(*x))
+ar.register_function("autograd", "gather", lambda x, indices: x[np.array(indices)])
+
+
+def _to_numpy_autograd(x):
+    if hasattr(x, "_value"):
+        # Catches the edge case where the data is an Autograd arraybox,
+        # which only occurs during backpropagation.
+        return x._value
+
+    return x.numpy()
+
+
+ar.register_function("autograd", "to_numpy", _to_numpy_autograd)
+
+
+def _scatter_element_add_autograd(tensor, index, value):
+    size = tensor.size
+    flat_index = _i("qml").numpy.ravel_multi_index(index, tensor.shape)
+    t = [0] * size
+    t[flat_index] = value
+    return tensor + _i("qml").numpy.array(t).reshape(tensor.shape)
+
+
+ar.register_function("autograd", "scatter_element_add", _scatter_element_add_autograd)
+
+
+def _take_autograd(tensor, indices, axis=None):
+    indices = _i("qml").numpy.asarray(indices)
+
+    if axis is None:
+        return tensor.flatten()[indices]
+
+    fancy_indices = [slice(None)] * axis + [indices]
+    return tensor[tuple(fancy_indices)]

```

Comment:
Seems to be missed by the tests

Reply:
This is super odd, because I checked locally and this is definitely being covered by this test: https://github.com/PennyLaneAI/pennylane/blob/cf28aa31537d555715b2270795656311e586748d/tests/math/test_functions.py#L923

Is there something I missed?

Reply:
oh, was leaning on the previous codecov status :thinking: it must be fine then! :)